util/file.py

```diff
@@ -830,3 +831,129 @@ def join_paired_fastq(input_fastqs, output_format='fastq', num_n=None):
+            return_obj["decompress_cmd"] = compressor + ["-dc"]
+            return_obj["compress_cmd"] = compressor + ["-c"]
+        elif filepath.endswith('.bz2'):
+            #compressor = 'lbzip2 {threads}'.format(threads="-n "+str(threads) if threads else "")
+            compressor = 'lbzip2 {threads}'.format(threads="-n "+str(threads) if threads else "").split()
+            return_obj["decompress_cmd"] = compressor + ["-dc"]
+            return_obj["compress_cmd"] = compressor + ["-c"]
+        elif filepath.endswith('.lz4'):
+            compressor = ['lz4']
+            return_obj["decompress_cmd"] = compressor +["-dc"]
+            return_obj["compress_cmd"] = compressor + ["-c"]
+        elif filepath.endswith('.tar'):
+            compressor = ['cat']
+            return_obj["decompress_cmd"] = compressor
+            return_obj["compress_cmd"] = compressor
+        else:
+            raise IOError("An input file of unknown type was provided: %s" % filepath)
+        return return_obj
+
+    def create_containing_dirs(path):
+        sub_path = os.path.dirname(path)
+        if not os.path.exists(sub_path) and sub_path != path:
+            create_containing_dirs(sub_path)
+        if not os.path.exists(path) and len(path):
+            os.mkdir(path)
+
+    class FileDiverter(object):
+        """
+            This reads bytes from a TarInfo file stream, writes them to a disk file
+            and returns the buffered bytes as they are read
+        """
+        def __init__(self, fileinfo, fileobj, written_mirror_file=None, extract_numeric_owner=False):
+            self.written_mirror_file = open(written_mirror_file,"wb")
+            self.fileinfo = fileinfo
+            self.fileobj = fileobj
+            self.extract_numeric_owner = extract_numeric_owner
+
+        def __del__(self):
+            self.written_mirror_file.flush()
+            self.written_mirror_file.close()
+
+            tar_in.chown(self.fileinfo, self.written_mirror_file.name, self.extract_numeric_owner)
+            if not self.fileinfo.issym():
+                tar_in.chmod(self.fileinfo, self.written_mirror_file.name)
+                tar_in.utime(self.fileinfo, self.written_mirror_file.name)
+
+        def read(self, size):
+            assert size is not None
+            
+            buf = self.fileobj.read(size)
+            self.written_mirror_file.write(buf)
+            return buf
+
+    if extract_to_disk_path and not os.path.isdir(extract_to_disk_path):
+        os.makedirs(extract_to_disk_path)
+
+    with open(out_compressed_tarball, "w") as outfile:
+        if out_compressed_tarball != "-":
+            out_compress_ps = subprocess.Popen(choose_compressor(out_compressed_tarball)["compress_cmd"], stdout=None if out_compressed_tarball == "-" else outfile, stdin=subprocess.PIPE)
+        else:
+            assert out_compressed_tarball != '-' or pipe_hint, "cannot autodetect compression for stdout output unless pipeHint provided"
```

Comment:
Isn't `out_compressed_tarball != '-'` just checked?

Reply:
Also may want to avoid using [assert](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) outside of unit test code.